### PR TITLE
Adding ray as a distributor

### PR DESCRIPTION
**Pull request by Vaishaal**
For running large jobs on AWS I found a couple problems with the spark backend. 

1. Spark runs things stage-wise by default so either you need to have a huge subjob_size and wait for the reader to finish before the job starts or you run stage by stage but then you are vulnerable to stragglers in each stage. In my experience I saw the the stragglers slowed down the pipeline by 2-10x depending how unlucky you got with a shard
2. when you are running on a non dedicated cluster spark doesn't handle autoscaling well so you basically need to pay for num_cpus x total runtime, whereas natively ray auto scales number of cpu nodes based on queue size and download speed. This is also nice when you get an error (as you often do for these large DL jobs), as you didn't waste money keeping cluster alive.
3. As of right now the ray aws utilities are much nicer to work with than the native spark ones.

This PR  just adds a ray distributor (with same interface as spark and multiprocessing distributors), and adds an example launch script, and a cluster_config.yaml file if people want to spin up their own AWS cluster.

Using this I was able to get over 200k/images second on a cluster of 100 m5.24xlarges consistently for 24 hours.

### CODE REVIEW

**Comment by Vaishaal** [comment]
@rom1504  can we merge this?

**Comment by rom1504** [comment]
Looks pretty good. Any way to add a test for this distributor ?
It would guarantee this would keep working with future changes

**Comment by jelech** [comment]
Hi, is there any plan to merge this PR?

**Comment by rom1504** [comment]
@jelech I want to but I would prefer we add some testing as otherwise it'll eventually get broken over time

**Comment by Vaishaal** [comment]
Let me do it!
________________________________

**Comment by rom1504** [comment]
I think we need this https://docs.ray.io/en/latest/cluster/vms/user-guides/launching-clusters/on-premises.html

**Comment by rom1504** [comment]
then adapt https://github.com/rom1504/img2dataset/blob/main/tests/test_main.py#L363

plus either automatically start a ray cluster if there is none in the main code 

or do it only in tests

File: img2dataset/distributor.py [hunk 1]
```diff
@@ -66,6 +66,29 @@ def run(gen):
...
+try:
+    import ray  # pylint: disable=import-outside-toplevel
+
+    @ray.remote
+    def ray_download(downloader, shards):
+        status, row = downloader(shards)
+        return status, row
+
+    def ray_distributor(processes_count, downloader, reader, _, max_shard_retry):  # type: ignore
+        # pylint: disable=unused-argument
```
Comment (rom1504) [inline img2dataset/distributor.py:78]: this should default to init ray locally

Comment (rom1504) [reply img2dataset/distributor.py:78]: actually this happens by default

**Comment by rom1504** [comment]
ok so I added tests

I'll rework the doc here a little bit and then we're good for merging

**Comment by Vaishaal** [comment]
Nice thanks so much!

**Comment by rom1504** [comment]
thanks for the PR!

**Pull request merged**